### PR TITLE
Bluetooth: Mesh: Move dfu mbt models into their own elements

### PR DESCRIPTION
**Pull request by m-alperen-sener**
- Tests: Bluetooth: Tester: Mesh DFUM and BLOB should have own elements
Moving the blob client, dfd server and dfu server to their own elements and increasing the CONFIG_BT_MESH_TX_SEG_MAX to 8 to fit new composition data into composition data page status message.
Standalone mesh blob client and DFU distributor/update server models requires one element and those elements only contain the main models and the models they extend to.
Referring to MshMBT_v1.0 Section 6.1: The BLOB Transfer Client model defines the messages listed in Table 6.1 , and requires one element: the BLOB Transfer Client Main element. The BLOB Transfer Client Main element contains the BLOB Transfer Client main model.
And referring to MshDFU_v1.0 Sections 6.1.1 and 6.2.1:
6.1.1
The Firmware Update Server model adds the state instances listed in Table 6.1 and Table 6.2 and the messages listed in Table 6.3 to the model it extends, and requires one element: the Firmware Update Main element. The Firmware Update Main element contains the Firmware Update Server main model and all the models that the main model extends.
6.2.1
The Firmware Distribution Server model adds the state instances listed in Table 6.7 and Table 6.8 and the messages listed in Table 6.9 to the model it extends, and requires one element: the Firmware Distribution Main element. The Firmware Distribution Main element contains the Firmware Distribution Server main model and all the models that the main model extends.

- Bluetooth: Mesh: Check that required models exists on the same element
Referring to MshDFU_v1.0 Sections 6.1.1, 6.2.1 and  7.1.1 model descriptions: DFU/DFD server/clients extend BLOB Transfer root models and DFD server requires Firmware Update Client on the same element. For this reason we need to make sure that those main models or root models exist on the same element. And also firmware update client can not be forced to be in the first element.

- For all model extention call return the error code in case of an error.

### CODE REVIEW

File: tests/bluetooth/tester/src/btp_mesh.c [hunk 1]
```diff
@@ -1114,15 +1123,6 @@ static const struct bt_mesh_model root_models_alt[] = {
 #if defined(CONFIG_BT_MESH_RPR_SRV)
 	BT_MESH_MODEL_RPR_SRV,
 #endif
-#if defined(CONFIG_BT_MESH_DFD_SRV)
```
Comment (alxelax) [inline tests/bluetooth/tester/src/btp_mesh.c:1117]: Why did you remove models from this list?
Adding them separately on the same element seems like misusing of `BT_MESH_ELEM`.
I'm surprised that such declaration works, since the macro was intended to declare model list for separate elements.

**Review by alxelax — CHANGES_REQUESTED** [review]

File: tests/bluetooth/tester/src/btp_mesh.c [hunk 2]
```diff
@@ -1165,10 +1166,28 @@ static const struct bt_mesh_model vnd_models[] = {
 
 static const struct bt_mesh_elem elements[] = {
 	BT_MESH_ELEM(0, root_models, vnd_models),
```
[the remark this answers is not in the document]
Comment (alxelax) [reply tests/bluetooth/tester/src/btp_mesh.c:1168]:
```
2.3.6 Models
...
This specification defines the following types of models:
•Base. A model that is extended by another model is called a base model.
•Root. A model is defined as a root model if it does not extend any other models. A root model may
serve as a base model for other models.
•Extending. A model that is extending another model is called an extending model.
•Corresponding. A model that has a correspond relationship with another model is called a
corresponding model.
•Main. A main model is a model at the end of an extension hierarchy that indicates an instance of a
functionality on a node. A main model is not extended by other models within the scope of the
functionality. A main model can be an extending model or a corresponding model.
```

File: tests/bluetooth/tester/src/btp_mesh.c (hunk 1 above)
Comment (m-alperen-sener) [reply tests/bluetooth/tester/src/btp_mesh.c:1117]: I did not get that; models are moved into different elements, they are not on the same element anymore!

 if you are talking about the first parameter of the BT_MESH_ELEM macro being 0, the first element is the location code not the element index on the list of element. http://www.bluetooth.com/specifications/assigned-numbers section 2.4.2.1
 
 If not so, I might need a bit more clarification on this.

File: subsys/bluetooth/mesh/dfu_cli.c [hunk 3]
```diff
@@ -965,11 +965,6 @@ static int dfu_cli_init(const struct bt_mesh_model *mod)
...
 	struct bt_mesh_dfu_cli *cli = mod->rt->user_data;
 
-	if (mod->rt->elem_idx != 0) {
-		LOG_ERR("DFU update client must be instantiated on first elem");
-		return -EINVAL;
-	}
-
 	cli->mod = mod;
 
 	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
```
[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: Here I only remove what is wrong, forcing dfu client to first element. The issue you mentioned is correct however is not in this commits scope.

also the helper macros for dfu and dfd currently covers most part, https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/mesh/dfu_cli.h#L48
but eventually it'd be nice have have check here when helpers are not used.

**Review by omkar3141 — DISMISSED** [review]
Looks fine to me, of course there are some old naming conventions issues, but it is up to you if you want to fix it.

File: tests/bluetooth/tester/src/btp_mesh.c (hunk 1 above)
Comment (alxelax) [reply tests/bluetooth/tester/src/btp_mesh.c:1117]: Yep, you are right. My mistake.

File: subsys/bluetooth/mesh/dfu_cli.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: Checking with zero element index implicitly checked that blob and dfu clients are on the same element. Since dfu client extends blob client, they must be on the same element. This is the mandatory statement is not checked anyhow. 
```
7.1.1 Description
The Firmware Update Client is a main model that extends the BLOB Transfer Client model
```

[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: Yes, Zephyr provides helper macros to avoid misconfiguration. However, there is no guarantee the macro was used.

File: tests/bluetooth/tester/src/btp_mesh.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply tests/bluetooth/tester/src/btp_mesh.c:1168]: Could you rename it?

[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply tests/bluetooth/tester/src/btp_mesh.c]: I changed it to primary_models_**

File: subsys/bluetooth/mesh/dfu_cli.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: I agree with you from the very beginning however currently this situation is not a bug, and adding check to models being in the same element depending on device role is kind of improvement and it is not scope of this PR. Also because this check should be added to other dfu/dfd/mbt models. Maybe it would be clearer and helpful if helpers named dfu_target, dfu_initiator, dfu_distributor and dfu_standalone_updater, because IMO currently helpers are not clear in purpose, dfd_srv, so what?

In brief that sounds like an imrovement more than a bug.

[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: What we need is to copy this 7 lines of code really:
https://github.com/zephyrproject-rtos/zephyr/blob/3364a35f05475a947dfd0bfc4f7b8984504caee4/subsys/bluetooth/mesh/priv_beacon_srv.c#L198-L204

[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: If we don't do this, absolutely anything can happen at the `bt_mesh_model_extend` call.

[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: then it is need to be done for every models which does not have this check?

[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/dfu_cli.c:970]: I think we spend more time arguing here than adding this check.

[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfu_cli.c:978]: applied the same check logic also for dfd_srv and dfu_srv to check required models are presented in the same element.

**Review by PavelVPV — DISMISSED** [review]

File: subsys/bluetooth/mesh/priv_beacon_srv.c [hunk 4]
```diff
@@ -206,7 +207,13 @@ static int priv_beacon_srv_init(const struct bt_mesh_model *mod)
 	priv_beacon_srv = mod;
 	mod->keys[0] = BT_MESH_KEY_DEV_LOCAL;
 
-	bt_mesh_model_extend(mod, config_srv);
+	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
```
Comment (alxelax) [inline subsys/bluetooth/mesh/priv_beacon_srv.c:210]: Please revert this change. This is not necessary. Sometime ago we had discussion about this and the decision was to move this dependency from source code to KConfig file. 
Please take a look: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/Kconfig#L1266-L1270

File: subsys/bluetooth/mesh/dfu_srv.c [hunk 5]
```diff
@@ -451,8 +452,20 @@ static int dfu_srv_init(const struct bt_mesh_model *mod)
...
 
+	const struct bt_mesh_model *blob_srv =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_BLOB_SRV);
+
+	if (blob_srv == NULL) {
+		LOG_ERR("Firmware Update Server cannot extend BLOB Transfer Server");
+		return -EINVAL;
+	}
+
 	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
```
Comment (alxelax) [inline subsys/bluetooth/mesh/dfu_srv.c:463]: Please remove `if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS))` condition for DFU server and client. I see KConfig dependencies are already correct.

**Review by alxelax — CHANGES_REQUESTED** [review]

**Review by alxelax — COMMENTED** [review]
LGTM at the moment. When you satisfy the Compliance checker, I'll approve it.

**Review by alxelax — DISMISSED** [review]

**Review by PavelVPV — DISMISSED** [review]

**Comment by PavelVPV** [comment]
@sjanc, could you please look at this PR?

File: subsys/bluetooth/mesh/dfd_srv.c [hunk 6]
```diff
@@ -925,12 +925,35 @@ const struct bt_mesh_blob_srv_cb _bt_mesh_dfd_srv_blob_cb = {
...
 
 	srv->mod = mod;
 
-	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
-		bt_mesh_model_extend(mod, srv->upload.blob.mod);
+	const struct bt_mesh_model *blob_srv =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_BLOB_SRV);
+
+	if (blob_srv == NULL) {
+		LOG_ERR("Firmware Distribution Server cannot extend BLOB Transfer Server");
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/dfd_srv.c:937]: Due to how we organize and discover models in the NCS Mesh, and how specification indicates they should arranged, I think it is better to shorten and reword this as:

```suggestion
		LOG_ERR("BLOB Transfer srv can't be on same element");
```

Long Debug string are not very useful and eat up the flash during development stage leading to misconceptions about footprint.

File: subsys/bluetooth/mesh/dfd_srv.c [hunk 7]
```diff
@@ -925,12 +925,35 @@ const struct bt_mesh_blob_srv_cb _bt_mesh_dfd_srv_blob_cb = {
...
+
+	/** BLOB client also shall be present on the same element, but it is already checked by
+	 * initiation of dfu client which we check here.
+	 */
+	const struct bt_mesh_model *dfu_cli =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_DFU_CLI);
+
+	if (dfu_cli == NULL) {
+		LOG_ERR("Firmware Distribution Server requires Firmware Update Client on the same "
+			"element");
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/dfd_srv.c:949]:
```suggestion
		LOG_ERR("Firmware Update cli needed on the element"
```
Long debug string are not very useful and eat up the flash during development stage leading to misconceptions about footprint.

File: subsys/bluetooth/mesh/dfu_cli.c [hunk 8]
```diff
@@ -963,17 +963,22 @@ const struct bt_mesh_model_op _bt_mesh_dfu_cli_op[] = {
...
 	struct bt_mesh_dfu_cli *cli = mod->rt->user_data;
+	cli->mod = mod;
+
+	const struct bt_mesh_model *blob_cli =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_BLOB_CLI);
 
-	if (mod->rt->elem_idx != 0) {
-		LOG_ERR("DFU update client must be instantiated on first elem");
+	if (blob_cli == NULL) {
+		LOG_ERR("Firmware Update Client cannot extend BLOB Transfer Client");
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/dfu_cli.c:974]:
```suggestion
		LOG_ERR("BLOB Transfer Client needed");
```

or 

```suggestion
		LOG_ERR("Missing BLOB cli");
```


Long debug string are not very useful and eat up the flash during development stage leading to misconceptions about footprint.

File: subsys/bluetooth/mesh/dfu_srv.c [hunk 9]
```diff
@@ -451,8 +452,18 @@ static int dfu_srv_init(const struct bt_mesh_model *mod)
 		return -EINVAL;
 	}
 
-	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
-		bt_mesh_model_extend(mod, srv->blob.mod);
+	const struct bt_mesh_model *blob_srv =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_BLOB_SRV);
+
+	if (blob_srv == NULL) {
+		LOG_ERR("Firmware Update Server cannot extend BLOB Transfer Server");
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/dfu_srv.c:459]:
```suggestion
		LOG_ERR("Missing BLOB srv");
```
Long debug string are not very useful and eat up the flash during development stage leading to misconceptions about footprint.

**Review by omkar3141 — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/dfd_srv.c (hunk 6 above)
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfd_srv.c:937]: Here the check is to make sure that Firmware Distribution Server is at the same element with BLOB Transfer server. Quite opposite of what you suggesting! Would you like me to update the commit and PR message if it is not clear enough?

**Comment by m-alperen-sener** [comment]
Tried to shorten error log messages without losing explanation for what the actual problem is. The key point is "not being on the same element", just saying "missing" might not give the message, or puzzle people. I am OK with removing all err logs anyway :+1:

**Comment by omkar3141** [comment]
> Tried to shorten error log messages without losing explanation for what the actual problem is. The key point is "not being on the same element", just saying "missing" might not give the message, or puzzle people. I am OK with removing all err logs anyway 👍

It is not necessary at all. If there is a problem, the log messages anyway print files names, etc. Even if MINIMAL is enabled, simply searching by string globally one can easily locate the log strings in code base. I feel, we shouldn't treat log messages as a way to educate users about what is the right thing to do. Instead, use them to actually print error codes or some useful information. This keeps application footprint low at development stage itself.

File: subsys/bluetooth/mesh/dfd_srv.c [hunk 10]
```diff
@@ -925,12 +925,34 @@ const struct bt_mesh_blob_srv_cb _bt_mesh_dfd_srv_blob_cb = {
...
 
 	srv->mod = mod;
 
-	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
-		bt_mesh_model_extend(mod, srv->upload.blob.mod);
+	const struct bt_mesh_model *blob_srv =
+		bt_mesh_model_find(bt_mesh_model_elem(mod), BT_MESH_MODEL_ID_BLOB_SRV);
+
+	if (blob_srv == NULL) {
+		LOG_ERR("Missing BLOB Srv on the same element.");
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/dfd_srv.c:937]: "on the same element" is unnecessary IMO. The code is quite self explanatory for anyone to understand what is wrong. With today's AI tools, with the context of "Missing BLOB Srv", any developer can understand what is missing. I will really prefer to keep things concise.

Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/dfd_srv.c:937]: OK; IMO the log messages are OK as it is.

> "on the same element" is unnecessary IMO.

Necessary IMO.

> The code is quite self explanatory for anyone to understand what is wrong.

In this regard removing log messages seems quite reasonable, I agree.

> With today's AI tools, with the context of "Missing BLOB Srv", any developer can understand what is missing. I will 
really prefer to keep things concise.

I don't understand this, can we decide on one thing, I feel like the discussion is beyond the scope, we can "discuss how a precise log message should be" somewhere else.

Comment (PavelVPV) [reply subsys/bluetooth/mesh/dfd_srv.c:937]: https://github.com/nrfconnect/sdk-nrf/blob/046312b6f9c008541e996cf603bf518a02e6503a/subsys/bluetooth/mesh/light_hsl_srv.c#L597

https://github.com/nrfconnect/sdk-nrf/blob/046312b6f9c008541e996cf603bf518a02e6503a/subsys/bluetooth/mesh/light_xyl_srv.c#L659

🙃

**Review by PavelVPV — APPROVED** [review]

File: subsys/bluetooth/mesh/dfd_srv.c (hunk 10 above)
Comment (omkar3141) [reply subsys/bluetooth/mesh/dfd_srv.c]: I reallyy dislike all of those now after seeing how footprint balloons.

**Review by omkar3141 — APPROVED** [review]
Thanks. Looks good to me.

**Review by alxelax — APPROVED** [review]

File: subsys/bluetooth/mesh/dfd_srv.c (hunk 10 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/dfd_srv.c]: Let's to not abuse this PR for preferences. I like long unique comments those describe what happens. Kind of disagree to shortened forms since they require additional investigation during debugging since this is not clear which one has been triggered.

Comment (alxelax) [reply subsys/bluetooth/mesh/dfd_srv.c]: > I reallyy dislike all of those now after seeing how footprint balloons.

This is an open source project. The message should be clear even for developers who didn't spend years for Mesh.

**Comment by kartben** [comment]
@sjanc PTAL

**Review by sjanc — APPROVED** [review]
